### PR TITLE
Fixes #108 - adding python code coverage

### LINKED ISSUE #108: Implement Python code coverage module support

### PR DESCRIPTION
**Pull request by ugocupcic**
Seems to be working: https://github.com/shadow-robot/ros-cpp-samples/pull/4

### CODE REVIEW

File: ansible/roles/ci/code_coverage/python/setup/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# Playbook for python code coverage
+
+- name: Set CATKIN_TEST_COVERAGE to True
+  set_fact:
+    python_code_coverage_options: "CATKIN_TEST_COVERAGE=1"
+
+- name: Install PIP coverage module
+  pip: name={{item}} extra_args='--upgrade'
+  with_items:
+     - coverage
```
Comment (AndriyPt) [inline ansible/roles/ci/code_coverage/python/setup/tasks/main.yml]: It is better to move this installation to ros_install script

File: ansible/docker_site.yml [hunk 2]
```diff
@@ -12,9 +12,11 @@
     - {role: ci/init/create_workspace, when: run_install, tags: ["create_workspace"]}
     - {role: ci/update_dependencies, when: run_build, tags: ["update_dependencies", "build", "unit_tests", "code_coverage", "codecov_tool"]}
```
Comment (AndriyPt) [inline ansible/docker_site.yml]: need to add python_code_coverage and cpp_code_coverage tages to update_dependencies, build, unit_tests and shippable_finalize roles

File: ansible/roles/ci/code_coverage/python/setup/tasks/main.yml (hunk 1 above)
Comment (ugocupcic) [inline ansible/roles/ci/code_coverage/python/setup/tasks/main.yml]: Here? https://github.com/shadow-robot/sr-build-tools/blob/master/ansible/roles/ci/init/install/tasks/main.yml#L15

**Comment by ugocupcic** [comment]
@AndriyPt OK everything should be fixed

**Comment by AndriyPt** [comment]
@ugocupcic have you tested both C++ and Python code coverage, are they working good together? Also it would be great if you refactor shippable/finalize and extract code from there to code_coverage/cpp/finalize . Use the same approach as you have used for Python (setup and finalize)

**Comment by ugocupcic** [comment]
@AndriyPt python seems to overwrite cpp coverage results. Looking into it.

File: ansible/docker_site.yml [hunk 3]
```diff
@@ -10,11 +10,18 @@
...
+    - {role: ci/code_coverage/python/setup, when: run_build, tags: ["python_code_coverage", "code_coverage", "codecov_tool"]}
+    - {role: ci/code_coverage/cpp/setup, when: run_build, tags: ["cpp_code_coverage", "code_coverage", "codecov_tool"]}
     - {role: ci/unit_tests, when: run_build, tags: ["unit_tests", "code_coverage", "codecov_tool"]}
     - {role: ci/code_style_check/python, when: run_build, tags: ["python_code_style_check", "code_style_check"]}
     - {role: ci/code_style_check/cpp, when: run_build, tags: ["cpp_code_style_check", "code_style_check"]}
-    - {role: ci/servers/shippable/finalize, when: run_build, tags: ["shippable","shippable_finalize"]}
+    - {role: ci/servers/shippable/finalize, when: run_build, tags: ["shippable","shippable_finalize",
+                                                                    "python_code_coverage", "cpp_code_coverage"]}
+    - {role: ci/code_coverage/python/finalize, when: run_build, tags: ["python_code_coverage", "code_coverage", "codecov_tool"]}
+    - {role: ci/code_coverage/cpp/finalize, when: run_build, tags: ["python_code_coverage", "code_coverage", "codecov_tool"]}
```
Comment (AndriyPt) [inline ansible/docker_site.yml]: tag should be cpp_code_coverage not python_code_coverage for ci/code_coverage/cpp/finalize

Comment (ugocupcic) [inline ansible/docker_site.yml]: fixed

File: ansible/roles/ci/code_coverage/cpp/finalize/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+---
+# Playbook for finalizing the cpp code coverage module
+
+- name: Run gcovr
+  shell: bash -c "gcovr -r {{ros_workspace}} --xml-pretty > {{code_coverage_results_dir}}/coverage.xml"
+    chdir={{code_coverage_results_dir}}
+  tags:
```
Comment (AndriyPt) [inline ansible/roles/ci/code_coverage/cpp/finalize/tasks/main.yml]: remove tags. They are no needed here. You have defined them in the docker_site.yml

Comment (ugocupcic) [inline ansible/roles/ci/code_coverage/cpp/finalize/tasks/main.yml]: fixed

File: ansible/docker_site.yml [hunk 5]
```diff
@@ -10,11 +10,18 @@
...
-    - {role: ci/build/catkin, when: run_build, tags: ["build", "unit_tests", "code_coverage", "codecov_tool"]}
-    - {role: ci/code_coverage/data, when: run_build, tags: ["code_coverage", "codecov_tool"]}
+    - {role: ci/update_dependencies, when: run_build, tags: ["update_dependencies", "build", "unit_tests",
+                                                             "code_coverage", "codecov_tool", "python_code_coverage",
+                                                             "cpp_code_coverage"]}
+    - {role: ci/build/catkin, when: run_build, tags: ["build", "unit_tests", "code_coverage", "codecov_tool",
+                                                      "python_code_coverage", "cpp_code_coverage"]}
+    - {role: ci/code_coverage/python/setup, when: run_build, tags: ["python_code_coverage", "code_coverage", "codecov_tool"]}
+    - {role: ci/code_coverage/cpp/setup, when: run_build, tags: ["cpp_code_coverage", "code_coverage", "codecov_tool"]}
     - {role: ci/unit_tests, when: run_build, tags: ["unit_tests", "code_coverage", "codecov_tool"]}
```
Comment (AndriyPt) [inline ansible/docker_site.yml]: add here "python_code_coverage", "cpp_code_coverage" as well

File: ansible/docker_site.yml [hunk 6]
```diff
@@ -10,11 +10,18 @@
...
     - {role: ci/code_style_check/python, when: run_build, tags: ["python_code_style_check", "code_style_check"]}
     - {role: ci/code_style_check/cpp, when: run_build, tags: ["cpp_code_style_check", "code_style_check"]}
-    - {role: ci/servers/shippable/finalize, when: run_build, tags: ["shippable","shippable_finalize"]}
+    - {role: ci/servers/shippable/finalize, when: run_build, tags: ["shippable","shippable_finalize",
```
Comment (AndriyPt) [inline ansible/docker_site.yml]: remove from here  "python_code_coverage", "cpp_code_coverage". We do not want to run this role if server is not shippable

File: ansible/docker_site.yml (hunk 5 above)
Comment (ugocupcic) [inline ansible/docker_site.yml]: fixed :)

**Comment by AndriyPt** [comment]
@ugocupcic Please fix path to C++ code coverage role. You have cpp/init and Python python/setup . So probably it should be cpp/setup as well :)

**Comment by ugocupcic** [comment]
@AndriyPt fixed

**Pull request merged**